### PR TITLE
x86 backend support for unsigned comparison

### PR DESCRIPTION
**Pull request by sviswa7**
x86 backend support for unsigned comparison

### CODE REVIEW

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 1]
```diff
@@ -9052,6 +9078,20 @@ void Assembler::evpcmpb(KRegister kdst, KRegister mask, XMMRegister nds, Address
   emit_int8((unsigned char)comparison);
 }
 
+void Assembler::evpcmpub(KRegister kdst, KRegister mask, XMMRegister nds, XMMRegister src,
```
Comment (PaulSandoz) [inline src/hotspot/cpu/x86/assembler_x86.cpp]: Do you think it worthwhile to combine `evpcmpb` and `evpcmpub`? If not then do they need to differ as much as they do?

File: src/hotspot/cpu/x86/x86.ad [hunk 2]
```diff
@@ -1226,6 +1226,11 @@ static inline Assembler::AvxVectorLen vector_length_encoding(const MachNode* use
   return vector_length_encoding(def);
 }
 
+static inline bool is_unsigned_booltest_pred(int bt) {
+  return  ((bt & BoolTest::unsigned_compare) == BoolTest::unsigned_compare);
```
Comment (PaulSandoz) [inline src/hotspot/cpu/x86/x86.ad:1230]: Defer to method on `BoolTest`?

File: src/hotspot/share/opto/subnode.hpp [hunk 3]
```diff
@@ -305,7 +305,8 @@ class CmpD3Node : public CmpDNode {
 // We pick the values as 3 bits; the low order 2 bits we compare against the
 // condition codes, the high bit flips the sense of the result.
 struct BoolTest {
-  enum mask { eq = 0, ne = 4, le = 5, ge = 7, lt = 3, gt = 1, overflow = 2, no_overflow = 6, never = 8, illegal = 9 };
+  enum mask { eq = 0, ne = 4, le = 5, ge = 7, lt = 3, gt = 1, overflow = 2, no_overflow = 6, never = 8, illegal = 9,
+                              unsigned_compare = 16 };
```
Comment (PaulSandoz) [inline src/hotspot/share/opto/subnode.hpp:309]: When the `unsigned_compare` is set how does it work with:

```c++
BoolTest::mask pred = (BoolTest::mask)cond->get_con();
```

given that there is no constant for an unsigned `lt` value etc? I am guessing that a `BoolTest::mask` value is an integer whose value need not be a member of the set of named constants?

File: src/hotspot/cpu/x86/assembler_x86.cpp (hunk 1 above)
Comment (sviswa7) [reply src/hotspot/cpu/x86/assembler_x86.cpp]: Yes, I can look into combining those two.

**Comment by sviswa7** [comment]
@PaulSandoz The signed/unsigned comparison instructions are merged in this update.

**Review by PaulSandoz — COMMENTED** [review]
> @PaulSandoz The signed/unsigned comparison instructions are merged in this update.

I think that looks much better: using an extra parameter for cases where signed/unsigned can be supported via the same instruction kind.

**Comment by PaulSandoz** [comment]
I am happy to merge this in as is, and iterate on it, if you are ok with that approach.

**Comment by sviswa7** [comment]
Sounds good, please go ahead and merge as is.

File: src/hotspot/cpu/x86/x86.ad (hunk 2 above)
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad:1230]: Yes, deferring to a method on BoolTest to check for unsigned sounds good.

File: src/hotspot/share/opto/subnode.hpp (hunk 3 above)
Comment (sviswa7) [reply src/hotspot/share/opto/subnode.hpp:309]: Yes, it is an integer and does not need to be in the set of named constant.

**Pull request merged**